Rxjs integration

tanadendere opened the pull request:
This pull request revamps the project to use RxJS observables for the network call.

tanadendere committed:
feat: :heavy_plus_sign: Installed RxJS

Added RxJS as a dev dependency

tanadendere committed:
refactor: :arrow_down: Fixing dependencies

Uninstalled and reinstalled some dependencies as dev dependencies and vice versa

tanadendere committed:
feat: :art: Separated static data

Removed static data from code and exported it from separate files

tanadendere committed:
feat: :sparkles: RxJS integration for network call

Reconfigured the code logic so it can easily accomodate an Observable and introduced the Observable

theadley commented on src/scripts/api.ts:
```diff
@@ -1,44 +1,30 @@
+    )
+  ),
+  catchError((error) => {
+    console.log("API Call has failed: ", error);
```
`console.error` level of verbosity/severity please

theadley commented on src/scripts/index.ts:
```diff
@@ -1,16 +1,18 @@
 // GLOBAL VARIABLES
-const map: Map = initilizeMap();
+let currentCity = majorCitiesData[0];
+console.log("Current city: ", currentCity.cityName);
```
No

theadley commented on src/scripts/index.ts:
```diff
@@ -32,17 +34,53 @@ const getMajorCities: NodeListOf<HTMLElement> =
-  city.addEventListener("click", function (e) {
-    getCityCoords(city.innerText);
+  city.addEventListener("click", function () {
+    console.log("this is running");
```
Nope

theadley commented on src/scripts/index.ts:
```diff
@@ -32,17 +34,53 @@ const getMajorCities: NodeListOf<HTMLElement> =
+      throw new Error(`No data for ${city.innerText}`);
+    }
+    weatherFetchRequest$.next(cityData);
+    subscribeToWeatherAPI(cityData);
```
Nope, you can't rebuild the marble run/hotwheels™ track every time you put a marble/car on the track

theadley commented on src/scripts/index.ts:
```diff
@@ -1,16 +1,18 @@
+
+// Loading current location on start up
+weatherFetchRequest$.next(currentCity);
+subscribeToWeatherAPI(currentCity);
```
Do this once here

theadley replied:
Not in the function, but the contents of the function, here, once, ever

theadley commented on src/scripts/index.ts:
```diff
@@ -32,17 +34,53 @@ const getMajorCities: NodeListOf<HTMLElement> =
+export function getCityCoords(city: string): ICityData {
+  for (const majorCity of majorCitiesData) {
+    if (majorCity.cityName === city) {
+      console.log(majorCity.cityName);
```
😭

theadley commented on src/scripts/index.ts:
```diff
@@ -32,17 +34,53 @@ const getMajorCities: NodeListOf<HTMLElement> =
+  // weatherFetchRequest$.next(cityData);
+
+  // console.log("hello 3");
+  console.log("in the subscribe to weather api function");
```
nooo

theadley commented on src/scripts/index.ts:
```diff
@@ -32,17 +34,53 @@ const getMajorCities: NodeListOf<HTMLElement> =
+
+  weatherResultSet$.subscribe((res) => {
+    console.log("subscribed");
+    console.log(res);
```
shhhhhh 🤫

theadley commented on src/scripts/index.ts:
```diff
@@ -32,17 +34,53 @@ const getMajorCities: NodeListOf<HTMLElement> =
+  // console.log("hello 4");
 }
+
+// export function callTheWeatherAPI(cityData: ICityData): void {
```
Don't PR unused code unless you can guarantee needing it in the future

theadley commented on src/scripts/map.ts:
```diff
@@ -1,25 +1,36 @@
+    .addTo(myMap)
+    .openPopup();
 
+  console.log("Map is initialised.");
```
👎🏻

theadley commented on src/scripts/map.ts:
```diff
@@ -29,21 +40,26 @@ export function mapListeners(
+  const mapButton$ = fromEvent(openMap, "click");
+  mapButton$.subscribe(() => {
+    mapDisplayToggle(mapDiv, sevenDayDiv);
+  });
```
I feel like you can use the traditional on click event handling

theadley commented on src/scripts/map.ts:
```diff
@@ -29,21 +40,26 @@ export function mapListeners(
   map.on("click", onMapClick);
 
-  function onMapClick(e) {
+  function onMapClick(e: any) {
```
Not any

Julianna-Venter replied:
`function onMapClick(e: { latlng: { lat: number; lng: number } }) `

theadley reviewed (changes requested):
Almost there. Just don't resetup your observable pipeline for every event

tanadendere committed:
fix: :bug: Removing the redundant subscriptions

Removed the subscription function and now only subscribe to the observable once

tanadendere committed:
refactor: :art: Changed parameter name for readability

Changed the parameter name of switchMap for better readability

tanadendere committed:
fix: :bug: Removing redundant subscription and observables in map function

theadley commented on src/scripts/index.ts:
```diff
@@ -32,17 +36,22 @@ const getMajorCities: NodeListOf<HTMLElement> =
+      return majorCity;
+    }
+  }
+  return {} as ICityData;
```
```ts
return majorCitiesData.find(cityObj => cityObj.cityName === city);
```

theadley commented on src/scripts/map.ts:
```diff
@@ -35,15 +42,15 @@ export function mapListeners(
+      cityName: "Selected Location",
     };
-    callTheWeatherAPI(cityData);
+    weatherFetchRequest$.next(cityData);
```
Here you want to call some provided callback function that requires an `ICityData` parameter, then `.next()` the observable from the main.ts

theadley commented on src/scripts/map.ts:
```diff
@@ -1,25 +1,32 @@
-import { map, latLng, marker, tileLayer, Map } from "leaflet";
-import { callTheWeatherAPI } from "./index.ts";
+import { map as createMap, latLng, marker, tileLayer, Map } from "leaflet";
+import { weatherFetchRequest$ } from "./api.ts";
```
Now the concerns are not separated

theadley reviewed (changes requested):
Separate out a bit more then you're sorted and SOLID

tanadendere committed:
refactor: :art: Separation of map concerns

tanadendere committed:
fix: :bug: More separation of map concerns

LesediMoko commented on src/scripts/index.ts:
```diff
@@ -22,7 +27,25 @@ const openMap = document.getElementById("map-button-float") as HTMLElement;
+    latitude: lat,
+    cityName: "Selected Location",
+  };
+  currentCity = cityData;
```
On line 43 you set the value of currentCity to a newly created cityData object but I do not see you using the currentCity variable again until you re-set its value to cityData again on line 64.

Is there a reason why you don't just reassign the value of currentCity. As in ` currentCity = {longitude: lng, latitude: lat, cityName: "Selected Location"} ` and just pass that into ` weatherFetchRequest$.next `

LesediMoko replied:
I just realised that the second cityData is not within the onMapClick function.
So you can just ignore the part about re-setting the value to cityData again, my bad 😅

LesediMoko commented on src/scripts/index.ts:
```diff
@@ -32,17 +55,17 @@ const getMajorCities: NodeListOf<HTMLElement> =
-  city.addEventListener("click", function (e) {
-    getCityCoords(city.innerText);
+  city.addEventListener("click", function () {
+    const cityData = getCityCoords(city.innerText);
```
Similar to my comment about line 43, would not be simpler to just have ` currentCity = getCityCoords(city.innerText); ` rather than creating const valuable that will just have its value assigned to currentCity just a few lines down?

tanadendere replied:
Yes, so I plan to do more verbose error handling if I get a chance... so currentCity is always guaranteed to have a value. 
So the separation is so that when I do the error handling, if cityData is null, I'll have a pop up to alert the user of the error but not crash the site.

LesediMoko replied:
Oh I see, you need currentCity to always have value as some sort of a backup. Understood 🤝

LesediMoko reviewed (commented):
Just something to consider about your usage of the currentCity variable ` GLOBAL VARIABLE `

theadley reviewed (approved):
There you go 👍🏻

LesediMoko reviewed (approved):
Carry on 👏
